core/src/main/scala/kafka/server/ConfigHandler.scala

```diff
@@ -42,29 +41,27 @@ trait ConfigHandler {
  */
 class TopicConfigHandler(private val logManager: LogManager, kafkaConfig: KafkaConfig) extends ConfigHandler with Logging {
 
-  def processConfigChanges(topic : String, topicConfig : Properties) {
-    val logs: mutable.Buffer[(TopicAndPartition, Log)] = logManager.logsByTopicPartition.toBuffer
-    val logsByTopic: Map[String, mutable.Buffer[Log]] = logs.groupBy{ case (topicAndPartition, log) => topicAndPartition.topic }
-            .mapValues{ case v: mutable.Buffer[(TopicAndPartition, Log)] => v.map(_._2) }
+  def processConfigChanges(topic: String, topicConfig: Properties) {
     // Validate the compatibility of message format version.
-    Option(topicConfig.getProperty(LogConfig.MessageFormatVersionProp)) match {
-      case Some(versionString) =>
-        if (!kafkaConfig.interBrokerProtocolVersion.onOrAfter(ApiVersion(versionString))) {
-          topicConfig.remove(LogConfig.MessageFormatVersionProp)
-          warn(s"Log configuration ${LogConfig.MessageFormatVersionProp} is ignored for $topic because $versionString " +
-            s"is not compatible with Kafka inter broker protocol version ${kafkaConfig.interBrokerProtocolVersion}")
-        }
-      case _ =>
+    val configNameToExclude = Option(topicConfig.getProperty(LogConfig.MessageFormatVersionProp)).flatMap { versionString =>
+      if (kafkaConfig.interBrokerProtocolVersion < ApiVersion(versionString)) {
+        warn(s"Log configuration ${LogConfig.MessageFormatVersionProp} is ignored for `$topic` because `$versionString` " +
+          s"is not compatible with Kafka inter-broker protocol version `${kafkaConfig.interBrokerProtocolVersion}`")
```

Comment:
Good point @junrao, changed it. I also looked for other similar instances of this and seems like they're all fixed now.